File: schemas/loyalty-rewards-program/drizzle/points_transactions.ts
// points_transactions: Immutable ledger of every point movement (earn, redeem, expire, adjust).
// See README.md for full design rationale.

import { pgTable, pgEnum, uuid, text, integer, boolean, timestamp, index } from "drizzle-orm/pg-core";
import { loyaltyMembers } from "./loyalty_members";
import { earningRules } from "./earning_rules";
import { promotions } from "./promotions";
import { rewardRedemptions } from "./reward_redemptions";

export const pointsTransactionType = pgEnum("points_transaction_type", ["earn", "redeem", "expire", "adjust", "bonus"]);

export const pointsTransactions = pgTable("points_transactions", {
    id: uuid("id").primaryKey().defaultRandom(),
    memberId: uuid("member_id").notNull().references(() => loyaltyMembers.id, { onDelete: "restrict" }),
    type: pointsTransactionType("type").notNull(),
    points: integer("points").notNull(),
    balanceAfter: integer("balance_after").notNull(),
    description: text("description"),
    sourceReferenceType: text("source_reference_type"),
    sourceReferenceId: text("source_reference_id"),
    earningRuleId: uuid("earning_rule_id").references(() => earningRules.id, { onDelete: "set null" }),
    promotionId: uuid("promotion_id").references(() => promotions.id, { onDelete: "set null" }),
    redemptionId: uuid("redemption_id").references(() => rewardRedemptions.id, { onDelete: "set null" }),
    expiresAt: timestamp("expires_at", { withTimezone: true }),
    isPending: boolean("is_pending").notNull().default(false),
    confirmedAt: timestamp("confirmed_at", { withTimezone: true }),
    createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
  }, (table) => [
    index("idx_points_transactions_member_id_created_at").on(table.memberId, table.createdAt),
    index("idx_points_transactions_type").on(table.type),
    index("idx_points_transactions_expires_at").on(table.expiresAt),
    index("idx_points_transactions_is_pending").on(table.isPending),
    index("idx_points_transactions_source_ref").on(table.sourceReferenceType, table.sourceReferenceId),
  ]
);

File: schemas/loyalty-rewards-program/spacetimedb/points_transactions.rs
// points_transactions: Immutable ledger of every point movement (earn, redeem, expire, adjust).
// See README.md for full design rationale.
use spacetimedb::Timestamp;

#[derive(SpacetimeType, Clone)]
pub enum PointsTransactionType {
    Earn,    // type: String
    Redeem,
    Expire,
    Adjust,
    Bonus,
}

#[spacetimedb::table(name = points_transactions, public)]
pub struct PointsTransaction {
    #[primary_key]
    pub id: String, // UUID

    #[index(btree)]
    pub member_id: String, // UUID, FK → loyalty_members.id (restrict delete)
    // Composite index: (member_id, created_at) — enforce in reducer logic
    #[index(btree)]
    pub r#type: PointsTransactionType,
    pub points: i32,
    pub balance_after: i32,
    pub description: Option<String>,
    #[index(btree)]
    pub source_reference_type: Option<String>,
    // Composite index: (source_reference_type, source_reference_id) — enforce in reducer logic
    pub source_reference_id: Option<String>,
    pub earning_rule_id: Option<String>, // UUID, FK → earning_rules.id (set null)
    pub promotion_id: Option<String>, // UUID, FK → promotions.id (set null)
    pub redemption_id: Option<String>, // UUID, FK → reward_redemptions.id (set null)
    #[index(btree)]
    pub expires_at: Option<Timestamp>,
    #[index(btree)]
    pub is_pending: bool, // default false
    pub confirmed_at: Option<Timestamp>,
    pub created_at: Timestamp,
}
